rapid7/metasploit-framework
# Create Module For CVE-2021-42321 - Exchange ChainedSerializationBinder DenyList Typo Authenticated RCE

This vulnerability allows remote attackers to execute arbitrary code on Exchange Server 2019 CU10 prior to Security Update 3, Exchange Server 2019 CU11 prior to Security Update 2, Exchange Server 2016 CU21 prior to Security Update 3, and Exchange Server 2016 CU22 prior to Security Update 2.

Note that authentication is required to exploit this vulnerability.

The specific flaw exists due to the fact that the blacklist for `ChainedSerializationBinder` had a typo whereby an 
entry was set as `System.Security.ClaimsPrincipal` instead of the proper value of `System.Security.Claims.ClaimsPrincipal`.

By leveraging this vulnerability, attackers can bypass the `ChainedSerializationBinder`'s deserialization blacklist
and execute code as `NT AUTHORITY\SYSTEM`.

Tested against Exchange Server 2019 CU11 SU0 on Windows Server 2019.


## To Implement n Touch Up On Before Final Land
- [x] Update exploit to randomize all numbers where possible.
- [x] Make the version affected description in exploit comply with standards
- [x] Remove extra comments where needed.
- [x] Determine if `ACCOUNT_LOCKOUT` is needed or not.

## To be implemented in a future PR
- [ ] Move the check code where possible into a mixin and rework the module to incorporate that.

## Verification

1. Set up a version of Windows Server 2019.
2. Download Exchange Server 2019 CU11 SU0 from https://download.microsoft.com/download/5/3/e/53e75dbd-ca33-496a-bd23-1d861feaa02a/ExchangeServer2019-x64-CU11.ISO
3. Follow the guide at https://petri.com/how-to-install-active-directory-in-windows-server-2019-server-manager to turn
the server into an AD server.
4. Mount the ISO and run `Setup.exe`. It should prompt you install .NET Framework, Visual Studio C++ Redistributables,
and Unified Communications Managed API. Install these and then reboot.
5. Follow https://www.nucleustechnologies.com/blog/step-by-step-guide-to-install-exchange-server-2019-part-1/ and
install the required features.
6. Keep running `Setup.exe` and installing extra dependencies as needed as per the links.
7. When you do get all dependencies installed, Exchange should give a button called `Install` which should no longer be
greyed out. Press this to install and accept any warnings that appear.
8. Go to https://*ip here*/owa/ and make sure you can see the Exchange Outlook login page.
